client_test.go

```diff
@@ -318,6 +319,71 @@ func TestClient_ResponseLogHook(t *testing.T) {
 	}
 }
 
+func TestClient_CheckRetry(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, "test_500_body", http.StatusInternalServerError)
+	}))
+	defer ts.Close()
+
+	client := NewClient()
+
+	retryErr := errors.New("retryError")
+	called := 0
+	client.CheckRetry = func(resp *http.Response, err error) (bool, error) {
+		if called < 1 {
+			called++
+			return DefaultRetryPolicy(resp, err)
+		}
+
+		return false, retryErr
+	}
+
+	// CheckRetry should return our retryErr value and stop the retry loop.
+	_, err := client.Get(ts.URL)
+
+	if called != 1 {
+		t.Fatal("we didn't retry the correct number of times")
+	}
+
+	if err == nil {
+		t.Fatal("expected error from request")
+	}
+
+	if err != retryErr {
+		t.Fatalf("Expected retryError, got:%v", err)
+	}
+}
+
+func TestClient_CheckRetryStop(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, "test_500_body", http.StatusInternalServerError)
+	}))
+	defer ts.Close()
+
+	client := NewClient()
+
+	// Verify that this stops retries on the first try, with no errors from the client.
+	called := 0
+	client.CheckRetry = func(resp *http.Response, err error) (bool, error) {
+		called++
+		return false, nil
+	}
+
+	_, err := client.Get(ts.URL)
+
+	if called == 0 {
+		t.Fatal("CheckRetry wasn't called")
+	}
+
+	if called > 1 {
```

Comment:
Let's just combine these to `called != 1`, and say like `expected 1 call to CheckRetry, got %d`